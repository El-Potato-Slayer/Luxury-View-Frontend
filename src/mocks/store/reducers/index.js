import { combineReducers } from 'redux';
import agentsReducer from './agentsReducer';
import appointmentsReducer from './appointmentsReducer';
import authReducer from './authReducer';
import propertiesReducer from './propertiesReducer';
import userReducer from './userReducer';
import filterReducer from './filterReducer';

export default combineReducers({
  agentsReducer, appointmentsReducer, authReducer, propertiesReducer, userReducer, filterReducer,
});
