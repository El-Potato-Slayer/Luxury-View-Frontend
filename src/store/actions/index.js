export const setAgents = (payload) => ({
  type: 'SET_AGENTS',
  payload,
});

export const setAppointments = (payload) => ({
  type: 'SET_APPOINTMENTS',
  payload,
});

export const setProperties = (payload) => ({
  type: 'SET_PROPERTIES',
  payload,
});

export const changeFilter = (name) => ({
  type: 'CHANGE_FILTER',
  name,
});
