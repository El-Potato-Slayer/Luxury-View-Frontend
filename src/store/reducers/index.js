import { combineReducers } from 'redux';
import agentsReducer from './agentsReducer';
import appointmentsReducer from './appointmentsReducer';
import propertiesReducer from './propertiesReducer';
import userReducer from './userReducer';
import filterReducer from './filterReducer';

export default combineReducers({
  agentsReducer, appointmentsReducer, propertiesReducer, userReducer, filterReducer,
});
