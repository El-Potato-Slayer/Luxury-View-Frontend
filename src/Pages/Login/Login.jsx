import axios from 'axios';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { setSelectedGuardRoute, setUser } from '../../Redux/actions/userActions';

function Login() {
  const history = useHistory();
  const { isLoggedIn, selectedGuardRoute } = useSelector((state) => state.userReducer);
  const [loginInfo, setloginInfo] = useState({
    username: '',
    password: '',
  });
  const [error, setError] = useState('');
  const dispatch = useDispatch();

  useEffect(() => {
    // Record path unatauthorized user tried to access before being asked to login
    if (!isLoggedIn && history.location.state) {
      dispatch(setSelectedGuardRoute(history.location.state.from.pathname));
    }
  }, [selectedGuardRoute]);

  const submitHandler = (e) => {
    setloginInfo({ ...loginInfo, [e.target.name]: e.target.value });
  };

  function postLogin() {
    axios.post('login', loginInfo)
      .then((res) => {
        localStorage.setItem('token', res.data.token);
        dispatch(setUser(res.data));
        if (selectedGuardRoute) {
          history.push(selectedGuardRoute);
        } else {
          history.push('/');
        }
      })
      .catch(() => {
        setError('Username or password is incorrect');
      });
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    postLogin();
  };

  return (
    <div className="form-wrapper" data-testid="login">
      <form className="form" onSubmit={handleSubmit}>
        {error}
        <h2 className="page-title">Login</h2>
        <fieldset>
          <input className="input-field" type="text" name="username" placeholder="Username" onChange={submitHandler} />
        </fieldset>
        <fieldset>
          <input className="input-field" type="password" name="password" placeholder="Password" onChange={submitHandler} />
        </fieldset>
        <button className="button info" type="submit">Submit</button>
      </form>
    </div>
  );
}

export default Login;
