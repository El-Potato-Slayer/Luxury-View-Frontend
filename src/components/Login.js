import axios from 'axios';
import { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { setAuth, setUser } from '../store/actions';

function Login() {
  const history = useHistory();

  if (history.location.state) {
    localStorage.setItem('redirectedLocation', history.location.state.from.pathname);
  }
  const [input, setInput] = useState({
    username: '',
    password: '',
  });
  const [error, setError] = useState('');
  const dispatch = useDispatch();

  const inputHandler = (e) => {
    setInput({ ...input, [e.target.name]: e.target.value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    axios.post('login', input)
      .then((res) => {
        localStorage.setItem('token', res.data.token);
        dispatch(setUser(res.data));
        dispatch(setAuth(true));
      })
      .catch(() => {
        setError('Username or password is incorrect');
      });
  };

  return (
    <form onSubmit={handleSubmit}>
      <h2>Login</h2>
      {error}
      <fieldset>
        <input type="text" name="username" placeholder="username" onChange={inputHandler} />
      </fieldset>
      <fieldset>
        <input type="password" name="password" placeholder="password" onChange={inputHandler} />
      </fieldset>
      <button type="submit">Submit</button>
    </form>
  );
}

export default Login;
