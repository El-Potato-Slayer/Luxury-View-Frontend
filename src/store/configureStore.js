import { applyMiddleware, createStore } from 'redux';
import ReduxThunk from 'redux-thunk';
import { fetchData } from './actions';
import { fetchAgentsFailure, fetchAgentsRequest, fetchAgentsSuccess } from './actions/agentActions';
import { fetchAppointmentsFailure, fetchAppointmentsRequest, fetchAppointmentsSuccess } from './actions/appointmentActions';
import { fetchPropertiesFailure, fetchPropertiesRequest, fetchPropertiesSuccess } from './actions/propertyActions';
import rootReducer from './reducers';

const store = createStore(rootReducer, applyMiddleware(ReduxThunk));

// store.subscribe(() => { console.log(store.getState()); });
store.dispatch(fetchData('properties', fetchPropertiesRequest, fetchPropertiesSuccess, fetchPropertiesFailure));
store.dispatch(fetchData('agents', fetchAgentsRequest, fetchAgentsSuccess, fetchAgentsFailure));
store.dispatch(fetchData('appointments', fetchAppointmentsRequest, fetchAppointmentsSuccess, fetchAppointmentsFailure, true));

export default store;
