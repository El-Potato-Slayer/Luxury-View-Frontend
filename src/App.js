import './App.css';
import {
  BrowserRouter, Redirect, Route, Switch,
} from 'react-router-dom';
import { connect, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Navbar from './components/Navbar';
import Home from './components/Home';
import { setAgents, setAuth, setProperties } from './store/actions';
import MansionsList from './containers/MansionsList';
import AgentsList from './containers/AgentsList';
import ProtectedRoute from './components/ProtectedRoute';
import AppointmentsList from './containers/AppointmentsList';
import Login from './components/Login';
import Mansion from './components/Mansion';

function App({ isAuth }) {
  const dispatch = useDispatch();
  let token;
  useEffect(() => {
    axios.get('http://localhost:3000/api/v1/properties')
      .then((resp) => {
        dispatch(setProperties(resp.data));
      });
    axios.get('http://localhost:3000/api/v1/agents')
      .then((resp) => {
        dispatch(setAgents(resp.data));
      });
    token = localStorage.getItem('token');
    if (token) {
      const tokenExpiration = JSON.parse(atob(token.split('.')[1])).exp;
      const dateNow = new Date();
      if (tokenExpiration > dateNow.getTime() / 1000) {
        dispatch(setAuth(true));
      }
    }
  }, []);
  return (
    <div data-testid="app" className="App">
      <BrowserRouter>
        <Navbar />
        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route exact path="/mansions">
            <MansionsList />
          </Route>
          <Route exact path="/agents">
            <AgentsList />
          </Route>
          <Route path="/login">
            {isAuth ? <Redirect to="/appointments" /> : <Login />}
          </Route>
          <Route path="/mansions/:id">
            <Mansion />
          </Route>
          <ProtectedRoute exact path="/appointments" component={AppointmentsList} />
        </Switch>
      </BrowserRouter>
    </div>
  );
}

const mapStateToProps = (state) => ({
  isAuth: state.authReducer.isAuth,
});

App.propTypes = {
  isAuth: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps)(App);
