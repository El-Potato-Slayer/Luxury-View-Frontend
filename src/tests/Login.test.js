import {
  cleanup, render, screen,
} from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import store from '../mocks/store/configureStore';
import App from '../App';
import Login from '../components/Login';

jest.mock('axios');
afterEach(cleanup);

test('Renders the Register component without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </App>
    </Provider>, div,
  );
});

test('Checks that title is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.getByTestId('login')).toHaveTextContent('Login');
});

test('Checks that unknown title is not rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.getByTestId('login')).not.toHaveTextContent('Jakey');
});

test('Checks that first name input is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Username')).toBeInTheDocument();
});

test('Checks that password input is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Password')).toBeInTheDocument();
});

test('Checks that unknown input is not rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Quake')).not.toBeInTheDocument();
});

test('Checks that submit button is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByText('Submit')).toBeInTheDocument();
});

it('matches snapshot', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
