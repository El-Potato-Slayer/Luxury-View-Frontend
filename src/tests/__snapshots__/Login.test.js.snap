// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="form-wrapper"
  data-testid="login"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    
    <h2
      className="page-title"
    >
      Login
    </h2>
    <fieldset>
      <input
        className="input-field"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        type="text"
      />
    </fieldset>
    <fieldset>
      <input
        className="input-field"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
    </fieldset>
    <button
      className="info-button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
